indra/newview/gltf/primitive.cpp

```diff
@@ -149,44 +324,131 @@ bool Primitive::prep(Asset& asset)
         }
     }
 
-    mVertexBuffer->setColorData(mColors.data());
+    mShaderVariant = 0;
 
     if (mNormals.empty())
     {
-        mNormals.resize(mPositions.size(), LLVector4a(0, 0, 1, 0));
+        mTangents.clear();
+
+        if (mMode == Mode::POINTS || mMode == Mode::LINES || mMode == Mode::LINE_LOOP || mMode == Mode::LINE_STRIP)
+        { //no normals and no surfaces, this primitive is unlit
+            mTangents.clear();
+            mShaderVariant |= LLGLSLShader::GLTFVariant::UNLIT;
+        }
+        else
+        {
+            // unroll into non-indexed array of flat shaded triangles
+            MikktMesh data;
+            if (!data.copy(this))
+            {
+                return false;
+            }
+
+            data.genNormals();
+            data.genTangents();
+            data.write(this);
+        }
+    }
+
+    bool unlit = (mShaderVariant & LLGLSLShader::GLTFVariant::UNLIT) != 0;
+
+    if (mTangents.empty() && !unlit)
+    { // NOTE: must be done last because tangent generation rewrites the other arrays
+        // adapted from usage of Mikktspace in llvolume.cpp
+        if (mMode == Mode::POINTS || mMode == Mode::LINES || mMode == Mode::LINE_LOOP || mMode == Mode::LINE_STRIP)
+        {
+            // for points and lines, just make sure tangent is perpendicular to normal
+            mTangents.resize(mNormals.size());
+            LLVector4a up(0.f, 0.f, 1.f, 0.f);
+            LLVector4a left(1.f, 0.f, 0.f, 0.f);
+            for (U32 i = 0; i < mNormals.size(); ++i)
+            {
+                if (fabsf(mNormals[i].getF32ptr()[2]) < 0.999f)
+                {
+                    mTangents[i] = up.cross3(mNormals[i]);
+                }
+                else
+                {
+                    mTangents[i] = left.cross3(mNormals[i]);
+                }
+
+                mTangents[i].getF32ptr()[3] = 1.f;
+            }
+        }
+        else
+        {
+            MikktMesh data;
+            if (!data.copy(this))
+            {
+                return false;
+            }
+
+            data.genTangents();
+            data.write(this);
+        }
     }
 
-    mVertexBuffer->setNormalData(mNormals.data());
 
-    if (mTangents.empty())
+    if (!unlit)
     {
-        // TODO: generate tangents if needed
-        mTangents.resize(mPositions.size(), LLVector4a(1, 0, 0, 1));
+        mask |= LLVertexBuffer::MAP_NORMAL;
+        mask |= LLVertexBuffer::MAP_TANGENT;
+    }
+
+    if (LLGLSLShader::sCurBoundShaderPtr == nullptr)
+    { // make sure a shader is bound to satisfy mVertexBuffer->setBuffer
+        gDebugProgram.bind();
     }
 
-    mVertexBuffer->setTangentData(mTangents.data());
+    mVertexBuffer = new LLVertexBuffer(mask);
+    mVertexBuffer->allocateBuffer(mPositions.size(), mIndexArray.size() * 2); // double the size of the index buffer for 32-bit indices
```

Comment:
It might be nice for there to be a flag in mVertexBuffer which says, "this VertexBuffer uses 32 bit indices"

Reply:
(not a blocker)